Added support for IKEA E2201 RODRET Dimmer.

yarafie opened the pull request:
Thank you for taking the time to work on a Pull Request. Your contribution is really appreciated! 🎉
Please don't delete any part of the template, since keeping the provided structure will help maintainers to review your work more rapidly.

Sections marked as * are required and need to be filled in.

Proposed change*
IKEA E2201 RODRET Dimmer - Initial Release -
Added support for IKEA E2201 RODRET Dimmer.

Checklist*
[Y] I followed sections of the [Contribution Guidelines (https://github.com/yarafie/awesome-ha-blueprints/blob/main/CONTRIBUTING.md) relevant to changes I'm proposing.
[Y] I properly tested proposed changes on my system and confirm that they are working as expected.
[Y] I formatted files with Prettier using the command npm run format before submitting my Pull Request.

yarafie commented:
@EPMatt I was playing around with github and messed things up.

So decided to just start over and do a new PR.

Foggy2 commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,768 @@
+                          - condition: template
+                            value_template: '{{ wait.trigger.idx is defined }}'
+                        then:
+                          - event: ahb_controller_event
+                            event_data:
+                              controller: '{{ controller_id }}'
+                              action: button_up_double
+                          - choose:
+                              - conditions: []
+                                sequence: !input action_button_up_double
```
```suggestion
              - choose:
                - conditions:
                  - condition: template
                    value_template: '{{ iif(action_up_button_double, true, false) }}'
                  sequence: !input action_button_up_double
                default:
                - event: ahb_controller_event
                  event_data:
                    controller: '{{ controller_id }}'
                    action: button_up_double
```

Foggy2 commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,768 @@
+  button_up_long_max_loop_repeats: !input button_up_long_max_loop_repeats
+  button_down_long_max_loop_repeats: !input button_down_long_max_loop_repeats
+  # Button Inputs for Virtual Double Press
+  helper_double_press_delay: !input helper_double_press_delay
```
```suggestion
  helper_double_press_delay: !input helper_double_press_delay
  action_up_button_double: !input action_button_up_double
```

Foggy2 reviewed (commented):
Hi @yarafie 

I was curious how automations work in HA so I have spent some time going through the YAML to understand it.

I think I have a solution for the actions configured on the controller blueprint conflicting with the actions configured in the linked hook.

Specifically, configuring custom actions on the controller will not function as expected. With both the hook and controller executing the configured actions for that event, instead of just the custom actions on the controller.

In my limited testing this seems to work well, but I am not very familiar with HA automations so have a look and let me know your thoughts.

yarafie commented:
Thanks. will take a look at it.
I also found a logic issue in the current blueprint itself, need to test it first has to do with when to call the hook events themselves, I think I was calling them more than they should be called,

yarafie commented:
> Hi @yarafie
> 
> I was curious how automations work in HA so I have spent some time going through the YAML to understand it.
> 
> I think I have a solution for the actions configured on the controller blueprint conflicting with the actions configured in the linked hook.
> 
> Specifically, configuring custom actions on the controller will not function as expected. With both the hook and controller executing the configured actions for that event, instead of just the custom actions on the controller.
> 
> In my limited testing this seems to work well, but I am not very familiar with HA automations so have a look and let me know your thoughts.

I haven't gotten to look at your proposel yet, but I just pushed a new version for the Rodret on my fork.

I think I fixed the hook and configured custom actions overlap.
I also added the ability to enable looping of custom actions on long press (Needs to be enabled otherwise default is not to loop)

If you are game to test it, test away.

Foggy2 commented:
@yarafie I have the updated blueprint setup on two remotes that are linked to lights. So far so good.

Have tested
- up (on)
- down (off)
- long press up (brightness up)
- long press down (brightness down)
- Double press up (color change)
- Double press down (color change)

With the exception of custom actions all seem to work well. I did not test any of the long press configuration options, only what was set by default.

Concerning the custom actions, as noted above, both the custom action and the standard hook action are triggered simultaneously which is undesirable.

Thanks!

yarafie commented:
> @yarafie I have the updated blueprint setup on two remotes that are linked to lights. So far so good.
> 
> Have tested
> 
> * up (on)
> * down (off)
> * long press up (brightness up)
> * long press down (brightness down)
> * Double press up (color change)
> * Double press down (color change)
> 
> With the exception of custom actions all seem to work well. I did not test any of the long press configuration options, only what was set by default.
> 
> Concerning the custom actions, as noted above, both the custom action and the standard hook action are triggered simultaneously which is undesirable.
> 
> Thanks!

Thanks for testing.

The way EPMatt has setup his platform is to fire both the hook actions and custom actions at the same time. That's the way it is designed.
When you say "With the exception of cystom actions" are they not working?"

Foggy2 commented:
> The way EPMatt has setup his platform is to fire both the hook actions and custom actions at the same time. That's the way it is designed.
> When you say "With the exception of cystom actions" are they not working?"

No, the custom actions are working correctly but they fire both the hook and the custom action at the same time.

I realize now, that perhaps I have made an incorrect assumption that this behavior is unintended.  My opinion is that if a custom action has been defined, then that should override the default action of the connected hook as multiple conflicting actions would cause undesired behaviors.

If it is intended that both actions are fired, then I would like to propose that another option be added to the UI that allows changing this behavior so that custom actions override the default hook.

esolitos commented:
> My opinion is that if a custom action has been defined, then that should override the default action of the connected hook as multiple conflicting actions would cause undesired behaviors.

I disagree: IMHO the correct behaviour is that all the Hooks should run.

> If it is intended that both actions are fired, then I would like to propose that another option be added to the UI that allows changing this behavior so that custom actions override the default hook.

While this could makes sense, it would be out of scope for this PR, as this should be implemented in all the controllers.

yarafie commented:
> > The way EPMatt has setup his platform is to fire both the hook actions and custom actions at the same time. That's the way it is designed.
> > When you say "With the exception of cystom actions" are they not working?"
> 
> No, the custom actions are working correctly but they fire both the hook and the custom action at the same time.
> 
> I realize now, that perhaps I have made an incorrect assumption that this behavior is unintended. My opinion is that if a custom action has been defined, then that should override the default action of the connected hook as multiple conflicting actions would cause undesired behaviors.
> 
Actually this gives more flexibility, but you really need to pay attention on how you set it up.
For example I have my RODRET Controlling two lamp lights using the Light hook so I csn turn off, turn on/,increase and decrease brightness using single and long presses.
The same RODRET I have set the double press custom action to turn on/off a closet light.

> If it is intended that both actions are fired, then I would like to propose that another option be added to the UI that allows changing this behavior so that custom actions override the default hook.

Would defeat the flexibility mentioned above.

yarafie commented:
> Holding the off button (O) just makes the brightness go down to 0 without stopping when the button is released.
> 
> I imported the new light and controller yaml files. Am I missing something?

Can you be more specific. zigbee2mqtt, zha.
Can you attach the automation setup.
If you know how to capture traces please do so to help in debugging.

EPMatt commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+            # Handle Down Double Press Event
+            then:
+              - choose:
+                  # wait for z2m/zha/dcz triggers for double press
+                  - conditions:
+                      - condition: trigger
+                        id:
+                          - z2m-button-down-short
+                          - zha-button-down-short
+                          - dcz-button-down-short
```
Same here

yarafie replied:
As Above

EPMatt commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                                            - zha-button-down-long
+                                            - dcz-button-down-long
+                                      sequence:
+                                        - wait_for_trigger:
+                                            # z2m triggers
+                                            - trigger: device
+                                              domain: mqtt
+                                              device_id: !input controller_device
+                                              type: action
+                                              subtype: brightness_stop
+                                            # zha triggers
+                                            - trigger: event
+                                              event_type: zha_event
+                                              event_data:
+                                                device_id: !input controller_device
+                                                command: stop
+                                                endpoint_id: 1
+                                                cluster_id: 8
+                                            # dcz triggers
+                                            - trigger: event
+                                              event_type: deconz_event
+                                              event_data:
+                                                device_id: !input controller_device
+                                                event: '2003'
```
Use the values from the `actions_mapping`  variable here, if possible. Otherwise, document that these are release events with a comment.

EPMatt commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                                            - zha-button-up-long
+                                            - dcz-button-up-long
+                                      sequence:
+                                        - wait_for_trigger:
+                                            # z2m triggers
+                                            - trigger: device
+                                              domain: mqtt
+                                              device_id: !input controller_device
+                                              type: action
+                                              subtype: brightness_stop
+                                            # zha triggers
+                                            - trigger: event
+                                              event_type: zha_event
+                                              event_data:
+                                                device_id: !input controller_device
+                                                command: stop_with_on_off
+                                                endpoint_id: 1
+                                                cluster_id: 8
+                                            # dcz triggers
+                                            - trigger: event
+                                              event_type: deconz_event
+                                              event_data:
+                                                device_id: !input controller_device
+                                                event: '1003'
```
Use the values from the actions_mapping variable here, if possible. Otherwise, document that these are release events with a comment.

yarafie replied:
As above

EPMatt commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                            - sequence:
+                                - choose:
+                                    - conditions:
+                                        - condition: trigger
+                                          id:
+                                            - z2m-button-down-long
+                                            - zha-button-down-long
+                                            - dcz-button-down-long
```
Same here. This seems to check that the trigger is any of the "long" events. But this was already checked previously (see line 670-675)

As a reference on the Trigger Condition:
https://www.home-assistant.io/docs/scripts/conditions/#trigger-condition

yarafie replied:
As above comments.

yarafie replied:
I do have a question related to long press hook events.
How does for example the light hook know that a long release was done?

Take as example the ikea_e1743, a button_long_release is sent but I do not see it in the light.yaml being captured.

EPMatt replied:
With `mode: restart`, the execution of the hook automation is halted as soon as a `release` event is fired by the controller, and piped to the hook.

EPMatt reviewed (changes requested):
Hey @yarafie,
first of all, thank you so much for your patience - it has been a hectic month on my side. 

I've requested a few changes in the blueprint + questions I believe we should address before merging.

Overall, I really love the cleaner logic for handling long press / release events, good job! 🔥  I believe we can migrate all existing controllers once we address the missing points here and publish this new blueprint.

EPMatt commented:
> > My opinion is that if a custom action has been defined, then that should override the default action of the connected hook as multiple conflicting actions would cause undesired behaviors.
> 
> I disagree: IMHO the correct behaviour is that all the Hooks should run.
> 
> > If it is intended that both actions are fired, then I would like to propose that another option be added to the UI that allows changing this behavior so that custom actions override the default hook.
> 
> While this could makes sense, it would be out of scope for this PR, as this should be implemented in all the controllers.

Hey there,
a clarification on the intended behavior: when a button is pressed , custom actions **and all** hooks associated with the controller should run.

yarafie commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                value_template: '{{ button_up_double_press_exposed }}'
+            # Handle Up Double Press Event
+            then:
+              - choose:
+                  # wait for z2m/zha/dcz triggers for double press
+                  - conditions:
+                      - condition: trigger
+                        id:
+                          - z2m-button-up-short
+                          - zha-button-up-short
+                          - dcz-button-up-short
+                    sequence:
```
I think you are correct here.
I will remove it and test.

yarafie commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                          - choose:
+                              - conditions: []
+                                sequence: !input action_button_up_short
+                # Not Up Double Handle Up Single Press Event
+                default:
+                  # fire the ahb hook event
+                  - event: ahb_controller_event
+                    event_data:
+                      controller: '{{ controller_id }}'
+                      action: button_up_short
+                  # run the custom action
+                  - choose:
+                      - conditions: []
+                        sequence: !input action_button_up_short
```
It is also probably not needed, I think I put it in as a final failsafe.
I will remove it and test.

yarafie commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,750 @@
+                            - sequence: !input action_button_up_long
+                            - sequence:
+                                - choose:
+                                    - conditions:
+                                        - condition: trigger
+                                          id:
+                                            - z2m-button-up-long
+                                            - zha-button-up-long
+                                            - dcz-button-up-long
```
As above

EPMatt commented on blueprints/controllers/ikea_e2201/ikea_e2201.yaml:
```diff
@@ -0,0 +1,684 @@
+    Hooks allow to easily create controller-based automations for interacting with lights, covers, media players, and potentially more.
+    See the list of [Hooks available for this controller](https://epmatt.github.io/awesome-ha-blueprints/docs/blueprints/controllers/ikea_e2201#available-hooks) for additional details.
+
+    A small note, avoid using both custom actions and Hooks in the same blueprint, otherwise strange things may occur.
+
```
```suggestion
```

In my view, this note would end up confusing the end user even more. We want to clarify the interaction between custom actions and hooks, potentially redesigning the feature to improve the UX.

See also recently opened discussion (feedback / suggestions wanted!) https://github.com/EPMatt/awesome-ha-blueprints/discussions/709

yarafie replied:
Removed from yaml and mdx. Forced pushed again.

EPMatt replied:
Quick note on force pushing: please only do it after rebasing the PR on top of `main` (e.g. in the event a new commit was added to `main`  while this PR was open). I think this time force pushing was not needed as this PR was already on top of the most recent `main` commit.

So in general, it's better to avoid force-pushing if not needed. :)
Oh, and when force pushing, [always `--force-with-lease`. ](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force)

Thanks again!

yarafie replied:
Noted. Just learning all the intricacies of git so I'm sure to screw up many many times. :)

yarafie replied:
What would have been the exact git commands that I should have done.

EPMatt replied:
My bad - I have accidentally force pushed my branch here and removed your last commit of my suggestion. Just reimplemented it again. Sorry for that :)

yarafie replied:
This is what I see now when I git switch to the branch

Switched to branch '20250220-yarafie-ikea-rodret-e2201-new-controller'
Your branch and 'upstream/main' have diverged,
and have 9 and 2 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

EPMatt replied:
Usually, the approach I take depends on the situation:

1. I am the only one concurrently working on a feature branch. Then, when an update occurs on the base branch.
    - `git fetch origin`
    - `git checkout <feature-branch>`
    - `git rebase origin/main`
    - `git push --force-with-lease`
    This allows to keep history clean on the PR, while ensuring all changes from the base branch are included. However, it might result in lost commits if multiple people collaborate and force push. Force push should be avoided at all costs in settings where multiple people actively work on the same branch - and if unavoidable, it's important everyone is informed before a force push. 
2. I am collaborating concurrently with more people on the same branch. Then, merging is the safest option:
    - `git fetch origin`
    - `git checkout <feature-branch>`
    - `git merge origin/main`
    - `git push`
    This adds a new merge commit in the PR branch history, cluttering the history a bit, but ensures no contributions are lost with force pushing. 

Probably, here 2. would have been the best option, but I  went with 1. by default. :)
Also, since we are squash merging into main when closing the PR, it's definitely ok to go with 2. for the next time.

yarafie replied:
The thing is I'm working off my fork which is the origin yours for me would be upstream.

My origin/main is slightly different from your upstream/main.
In order for me to properly test everything from start to finish My fork is almost a mirror image of your repo apart from having to change in most files references to epmatt/EPMatt to point to my fork yarafie/yarafie.

So I work directly in my fork, test everything.
Then when I'm comfortable I create a new branch based on your main (upstream/main) copy all files to that new branch. Update all references to epmatt/EPMatt. Update all changelogs, version numbers etc. Then do a PR.

Things get hairy only when other commits to main happen after my PR.

I hope the above makes sense.

EPMatt replied:
> apart from having to change in most files references to epmatt/EPMatt to point to my fork yarafie/yarafie.

@yarafie What are the tests you are running that you would not be able to perform without changing such references? It looks like that in order to test blueprint functionality, no reference updates are needed, and you could have a normal fork setup.

Trying to make things simpler here, so we can have an easier contribution flow. 👍🏻

yarafie replied:
If it was just testing the blueprint that woukd be easy.
I go through the whole testing through the website making sure the docs are fine.
Then importing the blueprint through the links then .... the whole user experience.

Sometimes just importing the blueprint points out errors which i can fix.

EPMatt replied:
@yarafie I see! I'd suggest you a different workflow:
- Create a Draft PR. This marks it as not ready for review.
- If linting is successful, the bot will post test links to updated blueprints as a PR comment.
- Import the blueprints in your system via the direct URL and test. 👍🏻 

There's not too much to worry about docs as long as you're copy-pasting the docs from another blueprint / doing minor edits - and in any case, CI has you covered in case of build failures. :)

Also, check out the new Development setup in the [Contributing Guidelines](https://github.com/EPMatt/awesome-ha-blueprints/blob/main/CONTRIBUTING.md). With the Devcontainer-based setup, you can easily spin up the Documentation Website locally by running `cd website && pnpm start`.

Btw just posting a few suggestions for you, but then I encourage you to follow the approach that suits you best. Feel free to let me know if you have any questions, doubts or you need any help while on this project. My goal here is to streamline your workflow and make it as easy as possible for you and others to contribute to the project and focus only on what matters - blueprints! :)

EPMatt reviewed (changes requested):
@yarafie love all the updates made to the controller logic in this blueprint - there is so much we can learn and apply to all the controllers. Thanks again for working on this!

Just a nit on docs I believe we should address before merging. Apart from that, we are good to go on my side!

Can't wait to see this and your other controller blueprints added to the collection 🔥
